beacon-chain/db/kv/finalized_block_roots.go

```diff
@@ -0,0 +1,108 @@
+package kv
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/boltdb/bolt"
+	"github.com/gogo/protobuf/proto"
+	"github.com/pkg/errors"
+	dbpb "github.com/prysmaticlabs/prysm/proto/beacon/db"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/featureconfig"
+	"github.com/prysmaticlabs/prysm/shared/traceutil"
+	"go.opencensus.io/trace"
+)
+
+var errMissingParentBlockInDatabase = errors.New("missing block in database")
+
+func updateFinalizedBlockRoots(ctx context.Context, tx *bolt.Tx, checkpoint *ethpb.Checkpoint) error {
+	if !featureconfig.Get().EnableFinalizedBlockRootIndex {
+		return nil
+	}
+
+	ctx, span := trace.StartSpan(ctx, "BeaconDB.updateFinalizedBlockRoots")
+	defer span.End()
+
+	bkt := tx.Bucket(finalizedBlockRootsIndexBucket)
+	blocks := tx.Bucket(blocksBucket)
+
+	root := checkpoint.Root
+	var previousRoot []byte
+	genesisRoot := blocks.Get(genesisBlockRootKey)
+
+	// Walk up the ancestry chain until we reach a block root present in the finalized block roots
+	// index bucket or genesis block root.
+	var walk = func() error {
+		for {
+			if bytes.Equal(root, genesisRoot) {
+				return nil
+			}
+
+			enc := blocks.Get(root)
+			if enc == nil {
+				traceutil.AnnotateError(span, errMissingParentBlockInDatabase)
+				return errMissingParentBlockInDatabase
+			}
+			block := &ethpb.BeaconBlock{}
+			if err := proto.Unmarshal(enc, block); err != nil {
+				traceutil.AnnotateError(span, err)
+				return err
+			}
+
+			container := &dbpb.FinalizedBlockRootContainer{
+				ParentRoot: block.ParentRoot,
+				ChildRoot:  previousRoot,
```

Comment:
The first time the walk is performed , `previousRoot` will be an empty byte slice

Reply:
That is expected, there is no finalized child root yet. We just learned of a new finalized block and this is the highest finalized block to exist so far.